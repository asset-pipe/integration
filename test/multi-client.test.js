'use strict';

const { resolve } = require;
const { spawn } = require('child_process');
const Client = require('@asset-pipe/client');
const buildServerUri = 'http://127.0.0.1:7200';
const client = new Client({ buildServerUri });
const supertest = require('supertest');
const request = supertest(buildServerUri);

const get = async url => {
    const { body } = await request
        .get(url)
        .set('Accept', 'application/json')
        .expect('Content-Type', /application\/json/)
        .expect(200);
    return body;
};

const localiseBodyPaths = body =>
    JSON.parse(
        JSON.stringify(body, null, 2).replace(
            /"file":\s".*\/asset-pipe\//g,
            '"file": "'
        )
    );

let server;
beforeAll(() => {
    jest.setTimeout(20000);
    return new Promise((resolve, reject) => {
        server = spawn('./node_modules/.bin/asset-pipe-server', [], {
            env: Object.assign({}, process.env, {
                NODE_ENV: 'development',
                PORT: 7200,
            }),
        });
        server.stdout.once('data', () => resolve());
        server.stderr.once('data', () => resolve());
        server.once('error', err => reject(err));
        server.once('close', () => resolve());
    });
});
afterAll(() => {
    server.kill();
});

test('Multiple clients upload js feeds to build server', async () => {
    expect.assertions(1);
    const js1 = resolve('../assets/a.js');
    const js2 = resolve('../assets/b.js');
    const js3 = resolve('../assets/c.js');
    const result = await Promise.all([
        client.uploadFeed([js1]),
        client.uploadFeed([js2]),
        client.uploadFeed([js3]),
    ]);
    expect(result).toMatchSnapshot();
});

test('Multiple clients upload css feeds to build server', async () => {
    expect.assertions(1);
    const css1 = resolve('../assets/a.css');
    const css2 = resolve('../assets/b.css');
    const css3 = resolve('../assets/c.css');
    const result = await Promise.all([
        client.uploadFeed([css1]),
        client.uploadFeed([css2]),
        client.uploadFeed([css3]),
    ]);
    expect(result).toMatchSnapshot();
});

test('Multiple clients get uploaded js feeds from build server', async () => {
    expect.assertions(1);
    const js1 = resolve('../assets/a.js');
    const js2 = resolve('../assets/b.js');
    const js3 = resolve('../assets/c.js');
    const [{ file: f1 }, { file: f2 }, { file: f3 }] = await Promise.all([
        client.uploadFeed([js1]),
        client.uploadFeed([js2]),
        client.uploadFeed([js3]),
    ]);
    const result = await Promise.all([
        get(`/feed/${f1}`),
        get(`/feed/${f2}`),
        get(`/feed/${f3}`),
    ]);
    expect(localiseBodyPaths(result)).toMatchSnapshot();
});

test('Multiple clients get uploaded css feeds from build server', async () => {
    expect.assertions(1);
    const css1 = resolve('../assets/a.css');
    const css2 = resolve('../assets/b.css');
    const css3 = resolve('../assets/c.css');
    const [{ file: f1 }, { file: f2 }, { file: f3 }] = await Promise.all([
        client.uploadFeed([css1]),
        client.uploadFeed([css2]),
        client.uploadFeed([css3]),
    ]);
    const result = await Promise.all([
        get(`/feed/${f1}`),
        get(`/feed/${f2}`),
        get(`/feed/${f3}`),
    ]);
    expect(result).toMatchSnapshot();
});

test('Client requests bundling of a js feed on build server', async () => {
    expect.assertions(1);
    const js1 = resolve('../assets/a.js');
    const js2 = resolve('../assets/b.js');
    const js3 = resolve('../assets/c.js');
    const [{ file: f1 }, { file: f2 }, { file: f3 }] = await Promise.all([
        client.uploadFeed([js1]),
        client.uploadFeed([js2]),
        client.uploadFeed([js3]),
    ]);
    const bundleResponse = await client.createRemoteBundle([f1, f2, f3], 'js');
    expect(bundleResponse).toMatchSnapshot();
});

test('Client requests bundling of a css feed on build server', async () => {
    expect.assertions(1);
    const css1 = resolve('../assets/a.css');
    const css2 = resolve('../assets/b.css');
    const css3 = resolve('../assets/c.css');
    const [{ file: f1 }, { file: f2 }, { file: f3 }] = await Promise.all([
        client.uploadFeed([css1]),
        client.uploadFeed([css2]),
        client.uploadFeed([css3]),
    ]);
    const bundleResponse = await client.createRemoteBundle([f1, f2, f3], 'css');
    expect(bundleResponse).toMatchSnapshot();
});

test('Client gets bundled js file from build server', async () => {
    expect.assertions(1);
    const js1 = resolve('../assets/a.js');
    const js2 = resolve('../assets/b.js');
    const js3 = resolve('../assets/c.js');
    const [{ file: f1 }, { file: f2 }, { file: f3 }] = await Promise.all([
        client.uploadFeed([js1]),
        client.uploadFeed([js2]),
        client.uploadFeed([js3]),
    ]);
    const bundleResponse = await client.createRemoteBundle([f2, f3, f1], 'js');
    const { text } = await request
        .get(`/bundle/${bundleResponse.file}`)
        .expect(200);
    expect(text).toMatchSnapshot();
});

test('Client gets bundled css file from build server', async () => {
    expect.assertions(1);
    const css1 = resolve('../assets/a.css');
    const css2 = resolve('../assets/b.css');
    const css3 = resolve('../assets/c.css');
    const [{ file: f1 }, { file: f2 }, { file: f3 }] = await Promise.all([
        client.uploadFeed([css1]),
        client.uploadFeed([css2]),
        client.uploadFeed([css3]),
    ]);
    const bundleResponse = await client.createRemoteBundle([f2, f3, f1], 'css');
    const { text } = await request
        .get(`/bundle/${bundleResponse.file}`)
        .expect(200);
    expect(text).toMatchSnapshot();
});

test('Bundled js feeds from build server are deduped', async () => {
    expect.assertions(1);
    const js1 = resolve('../assets/duped-before-uploading.js');
    const js2 = resolve('../assets/duped-before-uploading-2.js');
    const js3 = resolve('../assets/duped-before-uploading-3.js');
    const uploadResponse1 = await client.uploadFeed([js1]);
    const uploadResponse2 = await client.uploadFeed([js2]);
    const uploadResponse3 = await client.uploadFeed([js3]);
    const bundleResponse = await client.createRemoteBundle(
        [uploadResponse1.file, uploadResponse2.file, uploadResponse3.file],
        'js'
    );
    const { text } = await request
        .get(`/bundle/${bundleResponse.file}`)
        .expect(200);
    expect(text).toMatchSnapshot();
});
