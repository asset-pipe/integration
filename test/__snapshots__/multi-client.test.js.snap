// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client gets bundled css file from build server 1`] = `
"body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}

body {
    margin: 0;
    padding: 0;
}

.c-title {
    color: green;
}

body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

"
`;

exports[`Client gets bundled js file from build server 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({\\"3b0887ddfac98bb9c42ae30bdc580452e2798324bd16572e4635a7098e2139be\\":[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":\\"a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add\\"}],\\"56b86ab1072571a4efb08c9e5645d7ad7b442043bc29370df7212d1e7743e6ae\\":[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(5, 6);

},{\\"./d\\":\\"a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add\\"}],\\"82dd5eababe4c7ae7aeaa58e74cf83badd9d6066b08e6e6c1eb7a2935cf11d79\\":[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":\\"a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add\\"}],\\"a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add\\":[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}]},{},[\\"82dd5eababe4c7ae7aeaa58e74cf83badd9d6066b08e6e6c1eb7a2935cf11d79\\",\\"56b86ab1072571a4efb08c9e5645d7ad7b442043bc29370df7212d1e7743e6ae\\",\\"3b0887ddfac98bb9c42ae30bdc580452e2798324bd16572e4635a7098e2139be\\"]);
"
`;

exports[`Client requests bundling of a css feed on build server 1`] = `
Object {
  "file": "55960ccd31726d68f266bff255570763dbf3d98cf5c6dd8028f3f3d2430b5451.css",
  "uri": "http://127.0.0.1:7200/bundle/55960ccd31726d68f266bff255570763dbf3d98cf5c6dd8028f3f3d2430b5451.css",
}
`;

exports[`Client requests bundling of a js feed on build server 1`] = `
Object {
  "file": "7b5ad7930d053cd4c13de50de826dc5908ff39a5b4a790fcb6dd6d462b4bb70a.js",
  "uri": "http://127.0.0.1:7200/bundle/7b5ad7930d053cd4c13de50de826dc5908ff39a5b4a790fcb6dd6d462b4bb70a.js",
}
`;

exports[`Multiple clients get uploaded css feeds from build server 1`] = `
Array [
  Array [
    Object {
      "content": "body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}",
      "file": "integration/assets/a.css",
      "id": "1a54612eb24332ba5b5bbd272b577a4bc0ada16061c32328ca76dcce512917d0",
      "name": "integration",
      "version": "1.0.0",
    },
  ],
  Array [
    Object {
      "content": "body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}",
      "file": "integration/assets/b.css",
      "id": "e99eff35f3004172ec02c8cce828ca896802c19fbcecfa40c5f391a0762e45d9",
      "name": "integration",
      "version": "1.0.0",
    },
  ],
  Array [
    Object {
      "content": "body {
    margin: 0;
    padding: 0;
}

.c-title {
    color: green;
}",
      "file": "integration/assets/c.css",
      "id": "57f2dbebc30c76f889fe7578f84c6f34a8e79e217e80475e0a6861c871e580fa",
      "name": "integration",
      "version": "1.0.0",
    },
  ],
]
`;

exports[`Multiple clients get uploaded js feeds from build server 1`] = `
Array [
  Array [
    Object {
      "deps": Object {
        "./d": "a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add",
      },
      "entry": true,
      "expose": false,
      "file": "integration/assets/a.js",
      "id": "3b0887ddfac98bb9c42ae30bdc580452e2798324bd16572e4635a7098e2139be",
      "index": 1,
      "indexDeps": Object {
        "./d": 2,
      },
      "order": 0,
      "source": "'use strict';

const sum = require('./d');

sum(1, 2);
",
    },
    Object {
      "deps": Object {},
      "file": "integration/assets/d.js",
      "id": "a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add",
      "index": 2,
      "indexDeps": Object {},
      "source": "'use strict';

module.exports = function sum(a, b) {
    return a + b;
};
",
    },
  ],
  Array [
    Object {
      "deps": Object {
        "./d": "a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add",
      },
      "entry": true,
      "expose": false,
      "file": "integration/assets/b.js",
      "id": "82dd5eababe4c7ae7aeaa58e74cf83badd9d6066b08e6e6c1eb7a2935cf11d79",
      "index": 1,
      "indexDeps": Object {
        "./d": 2,
      },
      "order": 0,
      "source": "'use strict';

const sum = require('./d');

sum(3, 4);
",
    },
    Object {
      "deps": Object {},
      "file": "integration/assets/d.js",
      "id": "a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add",
      "index": 2,
      "indexDeps": Object {},
      "source": "'use strict';

module.exports = function sum(a, b) {
    return a + b;
};
",
    },
  ],
  Array [
    Object {
      "deps": Object {
        "./d": "a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add",
      },
      "entry": true,
      "expose": false,
      "file": "integration/assets/c.js",
      "id": "56b86ab1072571a4efb08c9e5645d7ad7b442043bc29370df7212d1e7743e6ae",
      "index": 1,
      "indexDeps": Object {
        "./d": 2,
      },
      "order": 0,
      "source": "'use strict';

const sum = require('./d');

sum(5, 6);
",
    },
    Object {
      "deps": Object {},
      "file": "integration/assets/d.js",
      "id": "a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add",
      "index": 2,
      "indexDeps": Object {},
      "source": "'use strict';

module.exports = function sum(a, b) {
    return a + b;
};
",
    },
  ],
]
`;

exports[`Multiple clients upload css feeds to build server 1`] = `
Array [
  Object {
    "file": "ba6124be6a330d945bc8b2e4798cde45320f4dc16c471f866e9fc991fb89d8fb.json",
    "id": "ba6124be6a330d945bc8b2e4798cde45320f4dc16c471f866e9fc991fb89d8fb",
    "uri": "http://127.0.0.1:7200/feed/ba6124be6a330d945bc8b2e4798cde45320f4dc16c471f866e9fc991fb89d8fb.json",
  },
  Object {
    "file": "515697d2559e49a972bdf09e0061435f845bef5c7236835ba675b3f2b1a95f47.json",
    "id": "515697d2559e49a972bdf09e0061435f845bef5c7236835ba675b3f2b1a95f47",
    "uri": "http://127.0.0.1:7200/feed/515697d2559e49a972bdf09e0061435f845bef5c7236835ba675b3f2b1a95f47.json",
  },
  Object {
    "file": "d0f49e024300c60cfa5f540061a18a228b863e81cfdbe0307d8d8e3096976ddb.json",
    "id": "d0f49e024300c60cfa5f540061a18a228b863e81cfdbe0307d8d8e3096976ddb",
    "uri": "http://127.0.0.1:7200/feed/d0f49e024300c60cfa5f540061a18a228b863e81cfdbe0307d8d8e3096976ddb.json",
  },
]
`;

exports[`Multiple clients upload js feeds to build server 1`] = `
Array [
  Object {
    "file": "6be8bba8c98ca6672c80c26488b88de531edd6f8ae4952756355b41a07e3fd73.json",
    "id": "6be8bba8c98ca6672c80c26488b88de531edd6f8ae4952756355b41a07e3fd73",
    "uri": "http://127.0.0.1:7200/feed/6be8bba8c98ca6672c80c26488b88de531edd6f8ae4952756355b41a07e3fd73.json",
  },
  Object {
    "file": "3e6959606e707cb36e7d538c72218bb72e4fe86c80f0666316135d933c77919e.json",
    "id": "3e6959606e707cb36e7d538c72218bb72e4fe86c80f0666316135d933c77919e",
    "uri": "http://127.0.0.1:7200/feed/3e6959606e707cb36e7d538c72218bb72e4fe86c80f0666316135d933c77919e.json",
  },
  Object {
    "file": "4c1a028075723f52e7c04e4d85bcbd78b7838a25a419730053ea0242ee8f76e8.json",
    "id": "4c1a028075723f52e7c04e4d85bcbd78b7838a25a419730053ea0242ee8f76e8",
    "uri": "http://127.0.0.1:7200/feed/4c1a028075723f52e7c04e4d85bcbd78b7838a25a419730053ea0242ee8f76e8.json",
  },
]
`;
